pkg/quarterdeck/invites.go

```diff
@@ -0,0 +1,203 @@
+package quarterdeck
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+
+	"github.com/cenkalti/backoff/v4"
+	"github.com/gin-gonic/gin"
+	"github.com/oklog/ulid/v2"
+	"github.com/rotationalio/ensign/pkg/quarterdeck/api/v1"
+	"github.com/rotationalio/ensign/pkg/quarterdeck/db"
+	"github.com/rotationalio/ensign/pkg/quarterdeck/db/models"
+	"github.com/rotationalio/ensign/pkg/quarterdeck/middleware"
+	perms "github.com/rotationalio/ensign/pkg/quarterdeck/permissions"
+	"github.com/rotationalio/ensign/pkg/quarterdeck/tokens"
+	"github.com/rotationalio/ensign/pkg/utils/sentry"
+	"github.com/rotationalio/ensign/pkg/utils/tasks"
+	"github.com/rotationalio/ensign/pkg/utils/ulids"
+)
+
+// InvitePreview returns details for a user invitation. This is a publicly accessible
+// endpoint because it has to be accessed by unauthenticated users without any claims.
+// Although the endpoint is read-only, it could return sensitive information about an
+// organization so this endpoint relies on the token being difficult to guess.
+func (s *Server) InvitePreview(c *gin.Context) {
+	var err error
+
+	// Get the token from the URL param
+	token := c.Param("token")
+
+	// Retrieve the invite from the database
+	var invite *models.UserInvitation
+	if invite, err = models.GetUserInvite(c.Request.Context(), token); err != nil {
+		if errors.Is(err, models.ErrNotFound) {
+			c.JSON(http.StatusBadRequest, api.ErrorResponse("invalid invitation"))
+			return
+		}
+
+		sentry.Error(c).Err(err).Msg("could not retrieve user invite")
+		c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not retrieve invitation"))
+		return
+	}
+
+	// Ensure the role is a recognized role
+	if !perms.IsRole(invite.Role) {
+		sentry.Warn(c).Str("role", invite.Role).Msg("invalid role for user invite")
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("invalid invitation"))
+		return
+	}
+
+	// Ensure the invite is valid and not expired
+	if err = invite.Validate(invite.Email); err != nil {
+		if errors.Is(err, db.ErrTokenExpired) {
+			c.JSON(http.StatusBadRequest, api.ErrorResponse("invitation has expired"))
+			return
+		}
+
+		sentry.Error(c).Err(err).Msg("bad user invite token")
+		c.JSON(http.StatusBadRequest, api.ErrorResponse("invalid invitation"))
+		return
+	}
+
+	// Attempt to lookup the user by their email address with their "default" organization
+	var user *models.User
+	if user, err = models.GetUserEmail(c.Request.Context(), invite.Email, ulid.ULID{}); err != nil {
+		if !errors.Is(err, models.ErrNotFound) {
+			sentry.Error(c).Err(err).Msg("could not retrieve user")
+			c.JSON(http.StatusInternalServerError, api.ErrorResponse("could not retrieve invitation"))
+			return
+		}
+
+		// Don't return an error to allow new users to create an account
```

Comment:
Stray comment for Preview? Should an error be returned on Preview?

Reply:
The idea is that the frontend needs to know whether or not the user exists to render the right form, and returning an error here would prevent the frontend from being able to render the content on the page.